File: [18] 4Sum/18.4-sum.java
import java.util.Arrays;

/*
 * @lc app=leetcode id=18 lang=java
 *
 * [18] 4Sum
 */

// @lc code=start
class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        // My solution: two-pointer
        List<List<Integer>> res = new ArrayList<>();
        if (nums.length < 4) return res;
        Arrays.sort(nums);
        if (target < nums[0]*4 || target > nums[nums.length-1]*4) return res;
        for (int i = 0; i < nums.length-3; i++) {
            if (nums[i]*4 > target) break;
            if (i != 0 && nums[i] == nums[i-1]) continue;
            for (int j = i+1; j < nums.length-2; j++) {
                if (nums[i] + nums[j]*3 > target) break;
                if (j != i+1 && nums[j] == nums[j-1]) continue;
                int lo = j+1, hi = nums.length-1;
                int sums = target-nums[i]-nums[j];
                while (lo < hi) {
                    if (nums[lo] + nums[hi] == sums) {
                        res.add(Arrays.asList(nums[i], nums[j], nums[lo], nums[hi]));
                        while (lo < hi && nums[lo] == nums[lo+1]) lo++;
                        while (lo < hi && nums[hi] == nums[hi-1]) hi--;
                        lo++;
                        hi--;
                    } else if (nums[lo] + nums[hi] < sums) lo++;
                    else hi--;
                }
            }
        }
        return res;
    }
}
// @lc code=end

File: [18] 4Sum/18.4-sum.cpp
/*
 * @lc app=leetcode id=18 lang=cpp
 *
 * [18] 4Sum
 */

// @lc code=start
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        // Hashset solution
        sort(nums.begin(), nums.end());
        return kSum(nums, target, 0, 4);
    }

    vector<vector<int>> kSum(vector<int>& nums, int target, int start, int k) {
        vector<vector<int>> res;
        // Note that nums.size() return unsigned int
        // So we can't use start > nums.size()-k, which can be negative
        if (start+k > nums.size() || nums[start]*k > target || nums.back()*k < target) return res;
        if (k == 2) return twoSum(nums, target, start);
        for (int i = start; i < nums.size(); i++) {
            if (i != start && nums[i] == nums[i-1]) continue;
            for (auto& set: kSum(nums, target-nums[i], i+1, k-1)) {
                res.push_back({nums[i]});
                res.back().insert(res.back().end(), set.begin(), set.end());
            }
        }
        return res;
    }

    vector<vector<int>> twoSum(vector<int>& nums, int target, int start) {
        vector<vector<int>> res;
        unordered_set<int> set;
        for (auto i = start; i < nums.size(); i++) {
            if (set.find(target - nums[i]) != set.end()) {
                if (res.empty() || res.back()[1] != nums[i]) {
                    res.push_back({target-nums[i], nums[i]});
                }
            }
            set.emplace(nums[i]);
        }
        return res;
    }
};
// @lc code=end

